Feat: Dynamic meta data for Twitter and Facebook cards

LulaV14 opened the pull request:
**Description**
I'm creating this PR to fix issue #1297 

Although is not completely done, I'll be working with @soueuls to finish this fix.

Bouncey commented on src/templates/Article.jsx:
```diff
@@ -22,6 +24,21 @@ function Article(props) {
+    .replace(/^\/([a-z0-9/-]+[^/])\/?$/i, '$1')
+    .split('/');
+
+  var metaTitle = '';
```
Why can't we use `title` from the `article`? On line 24

Bouncey replied:
Please declare with either `const` or `let`

soueuls replied:
I can't check the value of `title` right now (but if it's based on the page title or the front-matter specified one) we are afraid that it's going to be contextless is some cases.

Here is an [example](https://guide.freecodecamp.org/cplusplus/input_and_output) where the title is `Input and Output` without any mention of the context being C++.

We are going to work on a strategy to build it up from the url path, I think @LulaV14 is close to a solution.

Bouncey commented:
Changes to `meta.js` require a stop/start of gatsby as it effects the html template. This is maybe why you didn't see the changes immediately

soueuls commented:
@LulaV14 I made some research and I either need to make some pull request on your fork or you can add me as a collaborator with write access to it

LulaV14 commented:
@soueuls I've added you as a collaborator 😊 

Were you able to run the project so you can check my changes?
For example with this [url](https://guide.freecodecamp.org/java/loops/for-loop) `title` is "For Loop" and  `metaTitle` is "Java | Loops | For Loop" 

The problem is that even though by inspecting the elements you can see the tag 
`<meta content="Java | Loops | For Loop" property="og:title" data-react-helmet="true">`
using the [twitter card validator](https://cards-dev.twitter.com/validator) with ngrok I get an error.
This is the log I get:
```
INFO:  Page fetched successfully
INFO:  8 metatags were found
INFO:  twitter:card = summary tag found
ERROR: Possibly invalid value (if specified) (og:title)
ERROR: Possibly invalid value (if specified) (twitter:text:title)
ERROR: Possibly invalid value (if specified) (twitter:title)
```

soueuls commented:
@LulaV14 That's also the first time I am contributing with someone else on a repository I have no write access to, so I added your repository as upstream to work on top of your initial pull request. Made some changes and then made a pull request to freeCodeCamp. (We would need help to know if that's indeed the correct workflow)

My changes implement all the missing OG properties, fix the let/const.
I think what we are doing is wrong, we inject the title, description, and image dynamically but twitter reads the file content statically, therefore, the tags are not here yet.

Bouncey commented:
@soueuls that sounds about right, in development, everything about the article is dynamic. You would have to build and host the site to check that twitter is seeing everything properly.

QuincyLarson commented:
@LulaV14 This looks great. Were you able to address the changes @Bouncey requested? We're eager to get this live so these articles will look better when shared :)

LulaV14 commented:
Hi @QuincyLarson! Thanks, I've added the changes from @soueuls and deleted the cheerio implementation as requested. But I'm having trouble with the deployment to be able to test that everything is displaying properly.
What I'm doing is run `yarn run build` and it builds correctly but when I try to deploy to heroku it doesn't work using the node buildpack nor the react buildpack. Can you please help me with that?

Bouncey commented:
@LulaV14 we deploy with netlify, you should be able to host this branch there and test

Bouncey commented on src/templates/Article.jsx:
```diff
@@ -5,11 +5,54 @@ import Helmet from 'react-helmet';
+}
+
+function getOgDescription(html) {
+  const el = document.createElement('html');
```
We should be able to get an extract from graphql. Copy the page query at the end of the file and have a play around with the graphql server is which is available in development.

Bouncey commented on src/templates/Article.jsx:
```diff
@@ -5,11 +5,54 @@ import Helmet from 'react-helmet';
+}
+
+function getOgImage(html) {
+  const el = document.createElement('html');
```
I would query `markdownRemark` for this in graphql. `remark` converts the markdown file in to an AST. This might still be hanging around

LulaV14 commented:
@Bouncey I did try to use GraphQL as you requested but wasn't able to get anything from the html there. It may be done if we put fields for description and image just like it's done for the title in each post 
```
---
title: Tables
description: Tables description
image: https://somewebsite.com/images/tables.jpg
---
```

But since I wasn't successful with GraphQL what I did was to extract from the html variable the first paragraph and the first image using regex. I tested in local and also deployed to Netlify to be able to test it with the [twitter card validator](https://cards-dev.twitter.com/validator) and it works 😊  

Here is how it looks on Facebook:
<img width="503" alt="screen shot 2017-11-13 at 6 21 28 pm" src="https://user-images.githubusercontent.com/1399744/32759934-c2d33ffa-c8a0-11e7-9271-e7b42e17559e.png">

QuincyLarson commented:
@LulaV14 Thanks for your extreme patience. This looks great! This is a huge improvement over what we have now. I'm going to wait for @Bouncey to take a quick look at this and verify this is ready to merge.

Bouncey commented:
# LGTM

Thanks @LulaV14 and @soueuls :+1:

QuincyLarson commented:
@LulaV14 Thanks again for fixing this!

LulaV14 commented:
@QuincyLarson @Bouncey Thanks for letting me contribute to this awesome repo 😄 !! 
'Will be glad to be of any help again 😊
